eventuals/tcp.h

```diff
@@ -0,0 +1,583 @@
+#pragma once
+
+#include "eventuals/event-loop.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace eventuals {
+namespace ip {
+namespace tcp {
+
+////////////////////////////////////////////////////////////////////////
+
+enum class Protocol {
+  NOT_SET,
+  IPV4,
+  IPV6,
+};
+
+////////////////////////////////////////////////////////////////////////
+
+class Socket {
+ public:
+  Socket(EventLoop& loop = EventLoop::Default())
+    : loop_(loop),
+      socket_(loop.io_context()) {}
+
+  Socket(const Socket& that) = delete;
+  Socket(Socket&& that)
+    : loop_(that.loop_),
+      socket_(std::move(that.socket_)),
+      protocol_(that.protocol_) {}
+
+  Socket& operator=(const Socket& that) = delete;
+  Socket& operator=(Socket&& that) = delete;
+
+  auto Open(Protocol protocol) {
+    struct Data {
+      Socket* socket;
+      Protocol protocol;
+    };
+    Data data = {this, protocol};
+
+    return Eventual<void>()
+        .interruptible()
+        .context(data)
+        .start([](auto& context, auto& k, Interrupt::Handler& handler) {
+          asio::post(
+              context.socket->io_context(),
+              [context,
+               &k,
+               &handler]() {
+                if (handler.interrupt().Triggered()) {
+                  k.Stop();
+                } else {
+                  asio::error_code ec;
+
+                  switch (context.protocol) {
+                    case Protocol::IPV4:
+                      context.socket->underlying_handle().open(
+                          asio::ip::tcp::v4(),
+                          ec);
+                      break;
+                    case Protocol::IPV6:
+                      context.socket->underlying_handle().open(
+                          asio::ip::tcp::v6(),
+                          ec);
+                      break;
+                    default:
+                      k.Fail("Unknown protocol");
+                      return;
+                  }
+
+                  context.socket->protocol_ = context.protocol;
+
+                  if (!ec) {
```

Comment:
It reads really nice with `error`:
```suggestion
                  if (!error) {
```
😊